### PR TITLE
Merge 0.4.1

### PR DESCRIPTION
**Pull request by boulund**
Still some work remaining: mostly preparing the dev branch for release

### CODE REVIEW

File: README.md
Comment (jwdebelius) [inline README.md:5]: When emerged, this will need to be updated. Minor, but there

**Comment by jwdebelius** [comment]
I've only run through kraken and host removal; has anyone exorcized the pipeline fully?

File: README.md
Comment (boulund) [reply README.md:5]: Good call! 
I had a feeling I had forgotten about something, thanks for noticing!

**Comment by boulund** [comment]
@jwdebelius I haven't been able to implemented automated tests for all different variations of configurations. There is only a single automated test in place right now, and that's a basic syntax check and testing to see if Snakemake can build the entire DAG with all options activated. Suggestions for how to implement meaningful automated tests for (parts of) the pipeline are much appreciated!

**Comment by boulund** [comment]
This merge is put on hold until the development by @AroArz on adding AMRPlusPlus is finished and merged into `develop`.

File: rules/antibiotic_resistance/amrplusplus.smk
Comment (boulund) [inline rules/antibiotic_resistance/amrplusplus.smk:34]: perhaps better to use config["dbdir"] ?

**Comment by boulund** [comment]
I think we are getting ready to finally merge this now. I made some progress on other important stuff today so hopefully I have time to test this branch a bit tomorrow. Will report back here when I've taken it for a test spin

**Comment by AroArz** [comment]
How do you about waiting until I've sorted all the biobakery stuff? Will take a couple of weeks at max though.

**Comment by boulund** [comment]
I'll still take it for a spin tomorrow, then we can see. Perhaps we make this 4.1 release now and make a new release with the updated biobakery stuff afterwards.

**Comment by boulund** [comment]
Ok, so I've already discovered a couple of issues. I've been focusing on running with singularity as that is what I would like to focus on  as the most supported and reliable way of running StaG going forward. I'll put each thing I find in a separate comment following this one.

**Comment by boulund** [comment]
Encountered this error when trying to run the biobakery Singularity image. It seems metaphlan2.py is not available in PATH.

```
/bin/bash: metaphlan2.py: command not found
```

Investigating this led me to the following strange findings:

Snakemake pulled down shub://stag-mwc:biobakery as `b1997f03cfb69a5deb99cf2e329f8e18.simg`, I saw in the execution log that it was activated when running the metaphlan2 rules:
```
Activating singularity image /nfshome/home/boulund/stag_envs/b1997f03cfb69a5deb99cf2e329f8e18.simg
[Fri Jan 22 10:32:26 2021]
Error in rule metaphlan2:
    jobid: 0
    output: output_dir/metaphlan2/201_MiUdda__2017-B4-14375__V300056851_L2_128.bowtie2.bz2, output_dir/metaphlan2/2
    log: output_dir/logs/metaphlan2/201_MiUdda__2017-B4-14375__V300056851_L2_128.metaphlan2.stdout.log, output_dir/
    shell:

        metaphlan2.py             --input_type fastq             --nproc 5             --sample_id 201_MiUdda__2017

        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```

However, spawning a shell in that image it doesn't seem be the correct image:
```
(base) [boulund@c1hitachi10 stag_envs]$ singularity shell b1997f03cfb69a5deb99cf2e329f8e18.simg
Singularity> echo $PATH
/opt/conda/envs/stag-mwc-assembly/bin:/opt/conda/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/
sbin:/bin
Singularity> ls /opt/conda/envs/
stag-mwc-assembly
```
It actually looks like a copy of the assembly image. Not sure why it got mixed up. Will have to investigate further.

**Comment by boulund** [comment]
The other error I encountered was an argument typo in the call to metawrap in the assembly rule, the argument is just `--megahit`, not `--use-megahit` as I originally put in the code. Not sure why I haven't spotted that before.

**Comment by boulund** [comment]
At least I've verified that the email on workflow completion/failure is working! Got an email about the failed run :D

**Comment by boulund** [comment]
### AMRPlusPlus findings
I noted that the `align_to_amr` rule is using `bwa mem` to perform alignments with a fixed number of threads. This should be made configurable like other rules that uses multithreading. Can you take care of that @AroArz ?

Please also add those to the cluster profiles as well if you can.

**Comment by AroArz** [comment]
> ### AMRPlusPlus findings
> I noted that the `align_to_amr` rule is using `bwa mem` to perform alignments with a fixed number of threads. This should be made configurable like other rules that uses multithreading. Can you take care of that @AroArz ?
> 
> Please also add those to the cluster profiles as well if you can.

I will fix this

**Comment by boulund** [comment]
Getting some errors in checkm of the binning step:
```
*******************************************************************************
 [CheckM - tree] Placing bins in reference genome tree.
*******************************************************************************

  Identifying marker genes in 2 bins with 20 threads:
Process SyncManager-1:
Traceback (most recent call last):
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/process.py", line 267, in _bootstrap
    self.run()
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/managers.py", line 550, in _run_server
    server = cls._Server(registry, address, authkey, serializer)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/managers.py", line 162, in __init__
    self.listener = Listener(address=address, backlog=16)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/connection.py", line 132, in __init__
    self._listener = SocketListener(address, family, backlog)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/connection.py", line 256, in __init__
    self._socket.bind(address)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
error: AF_UNIX path too long
Traceback (most recent call last):
  File "/opt/conda/envs/stag-mwc-assembly/bin/checkm", line 708, in <module>
    checkmParser.parseOptions(args)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/site-packages/checkm/main.py", line 1251, in parseOptions
    self.tree(options)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/site-packages/checkm/main.py", line 133, in tree
    options.bCalledGenes)
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/site-packages/checkm/markerGeneFinder.py", line 67, in find
    binIdToModels = mp.Manager().dict()
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/__init__.py", line 99, in Manager
    m.start()
  File "/opt/conda/envs/stag-mwc-assembly/lib/python2.7/multiprocessing/managers.py", line 528, in start
    self._address = reader.recv()
EOFError
```

**Comment by boulund** [comment]
I'm starting to get pretty happy with how StaG works in the develop branch now. Anything else important that you want to fix before we merge @AroArz ?

**Comment by AroArz** [comment]
I also experienced the singularity cache mix-up when running HUMAnN2. Snakemake downloaded the correct image but failed to activate the image because of cache inconsistency. I instead pointed directly to the image in `.snakemake/singularity/` which solved the issue. 

I have other updates in mind but we can take them together with the Biobakery update e.g. figuring out a solution for the huge output files from AMR++.

**Comment by boulund** [comment]
> the huge output files from AMR++.
Do you mean HUMAnN? Regardless. I think it sounds good.

I will make one last read-through of the docs to see if there's anything that needs updating in the develop branch, then I think it's finally time to merge 0.4.1! 🎉

**Review by AroArz — APPROVED** [review]

**Comment by AroArz** [comment]
We had a discussion a while ago about the output from the `run_resistome` rule from AMR++ which created intermediary files only used by a couple of rules but which occupied a lot of disk space. I can't remember which files these were but I think it might be worth looking over in the next update. I remember Yue bringing up this issue when she used the pipeline.

**Comment by boulund** [comment]
Yes, now I remember. It was probably the large sam files? We can probably change those to be `temp()` in our next review of those sections of the workflow.

**Comment by AroArz** [comment]
Most likely the sam files. Sounds good!

**Pull request merged**
